apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -87,76 +101,148 @@ export function positionBlocksOnWorkspace(workspace, blockOrderMap) {
     false
   );
 
-  orderedBlocksSetupFirst.forEach(block => {
-    positionBlockWithCursor(block, cursor);
+  adjustBlockPositions(orderedBlocksSetupFirst, workspace);
+}
+
+/**
+ * Positions blocks with any mix of user-defined positions and default positions (including all of one or the other)
+ * Such that none of the blocks overlap with each other
+ * @param {Array<Blockly.Block>} blocks - The blocks to position
+ * @param {Blockly.Workspace} workspace - The current Blockly workspace
+ */
+function adjustBlockPositions(blocks, workspace) {
+  let existingColliders = [];
+  let blocksToPlace = [];
+  blocks.forEach(block => {
+    if (isBlockLocationUnset(block)) {
+      blocksToPlace.push(block);
+    } else {
+      insertCollider(existingColliders, getCollider(block));
+    }
+  });
+
+  let y = WORKSPACE_PADDING;
+  blocksToPlace.forEach(block => {
```

Comment:
I think we want to move the declaration for `y` above into this loop.

For each block we want to place, we should assume it might be small enough to fit into a space that a previous block could not. For example, compare how these two clips handle the `NEW` function definition block. Without resetting `y` for each new block, we mistakenly assume it needs to start no higher than the previously placed block.


https://github.com/code-dot-org/code-dot-org/assets/43474485/97b48209-f13f-4b67-97db-0585e5ce246b


https://github.com/code-dot-org/code-dot-org/assets/43474485/cb3bdfe9-cefe-4e00-9c93-e19dcfe65379